### ISSUE TITLE
Shielded transaction aggregates

### ISSUE THREAD

**Issue by daira** [zcash/zcash#4946]
## Is your feature request related to a problem? Please describe.

If Zcash achieves significantly higher usage, then the overall verification and transaction bandwidth cost will become a problem. In other cryptocurrencies this has led to high and less predictable fees, increases in latency of transactions being mined, and obstacles to adoption &mdash; both immediate and due to concerns about future scalability.

## Describe the solution you'd like
A "shielded transaction aggregate" is a compact representation of a set of non-conflicting transactions. Aggregates and individual transactions can be combined off-chain, and then included on-chain with much lower bandwidth usage and verification cost.
This approach is essentially a "rollup" in the terminology Vitalik uses [here](https://vitalik.ca/general/2021/01/05/rollup.html), but note that most other rollup protocols do not support transaction graph privacy, and most do not use recursive proofs.

An aggregate consists of:
* a single Halo 2 proof which recursively validates a tree of transactions;
* a single Merkle root to a subtree of note commitments, to be added to the global note commitment tree;
* a list of nullifiers, to be added to the global nullifier set;
* a list of note ciphertexts;
  * optionally, note ciphertexts can be split from memo ciphertexts.

The proposal is parameterized by a value N which is simultaneously, for any aggregate:
* the maximum size of the commitment subtree;
* the maximum number of nullifiers;
* the maximum number of note ciphertexts.

For simplicity and reduction of attack surface, only fully shielded transactions spending Pollard notes to Pollard addresses are supported. No new shielded pool or address type is introduced, and no transparent or TZE functionality is supported for aggregated transactions. This does not introduce any immediate compatibility problem because single transactions can be mixed with aggregates in a block (see "Inclusion in blocks" below).

### Scalability

This doesn't by itself change the _asymptotic_ behaviour of the protocol:
* the bandwidth of transaction broadcast is asymptotically the same because a tx aggregate still contains all the nullifiers of spent notes, and some of the note ciphertexts;
* scanning of the remaining note ciphertexts is still necessary.

However, it can significantly reduce the concrete bandwidth use and verification cost.

#### On-chain bandwidth

The size of a nullifier is 32 bytes, and the total size of a note+outgoing ciphertext including epk and memo is 692 bytes. If the memo is split into a different ciphertext, then the non-memo+outgoing part would be ~180 bytes. This compares with 384 + 948 = 1332 bytes for a Spend description and Output description in Sapling, and it is likely that this will be somewhat greater in Pollard.

I estimate, very provisionally, that the size of a fully shielded Pollard transaction with two "actions" (up to two spends and two outputs, one for change; no memo for the change) will be 4767 bytes. The corresponding Sapling transaction would be 2756 bytes.

Define the *footprint* of a transaction to be the size of its fields that are included on chain, i.e. nullifiers and note ciphertexts. The above Pollard transaction, included in an aggregate, would have a footprint of 32*2 + 692 + 180 = 936 bytes, if both outputs are sent in-band.

So, the potential reduction of transaction bandwidth *relative to Sapling* is in the range roughly ~66% (all outputs in-band) to ~97.7% (all outputs out of band). *Relative to Pollard*, it is in the range roughly ~80% (all outputs in-band) to ~98.6% (all outputs out of band).

TODO: bandwidth calculations for light clients.

#### Verification cost

Furthermore, the main bottleneck to increasing transaction capacity is arguably full-node verification cost rather than on-chain bandwidth. An aggregate of up to N transactions can be verified faster than a single Pollard transaction (because verifying the recursive proof is also verifying signatures, which have to be verified separately for single Pollard transactions).

This proposal is also a pragmatic step toward a succinct block chain, if the latter turns out to be needed. Transaction aggregation would be a necessary part of a succinct block chain design in any case. It is complementary to sharding or other potential scaling approaches. It's also complementary to using a more succinct nullifier non-membership accumulator.

### Fees

Define the *footprint* of a transaction to be the size of its fields that are included on chain, i.e. nullifiers and note ciphertexts.
A fee is paid by each transaction creator to the eventual miner, at a fixed rate per footprint byte. That is, an aggregate pays a miner fee that is the sum of the miner fees from the aggregated transactions.

It would be possible to allow the aggregators to also take some fraction of the fees. However this is tricky to define because there can be multiple levels of aggregation by different parties, and so I'm omitting it from this initial proposal.

Note that aggregators may have external incentives to perform their task even without fees. For example, I expect it would be common for exchanges to aggregate their users' transactions. In that case the exchange already typically takes a fee, and does not need another one to cover the aggregation cost, which is part of the service they are providing to their users.

If we decide that full decentralization of proving is a requirement, then we should look at other proving pool designs such as Mina's "SNARKetplace".

### Receipts

A tx creator can tell that a spend (and therefore the transaction that included it) has been included in the block chain, by scanning all nullifiers.

It may be desirable for a tx aggregator to give a proof to a tx creator that it has accepted the tx and promises to submit it to the block chain in some aggregate. This is feasible as an extension to this design. The tx creator SHOULD set an expiry height so that it knows when to resubmit the tx (either to another aggregator, or directly as a single tx) in case it has not been included.

### Output discovery

Including a note ciphertext for an output is optional. The proof ensures that all note ciphertexts that were intended to be included by tx creators are actually included (using a Merkle tree over the ciphertexts).

This is intended to allow a _gradual_ switch from broadcast note ciphertexts on the block chain, to point-to-point communication of notes over privacy-protecting or local networks.

### Inclusion in blocks

An aggregate would be represented in a block as a different transaction version. Single transactions and aggregates compete for block space. There might be a case for reserving some portion of the block space that is guaranteed to be available for aggregates.

## Alternatives you've considered

My presentations at Zcon1 and [at the Amsterdam zkproof event](https://www.youtube.com/watch?v=RN82M_ShLtw) describe the most comprehensive straw design for scaling in Zcash that had previously been proposed. This design addresses a lot of issues that are arguably not necessary to solve in an initial deployment of scalability improvement. It included transaction aggregation as one of the ideas, but this ticket lays out that part in more detail, and with some simplifications.

**Comment by nathan-at-least** [comment]
This sounds like an excellent building block for further capacity improvements. Nice work! :-)

### Miner Aggregation Question

Question: Is it possible for a miner to perform aggregation over their entire mempool? This would allow the current p2p broadcast mechanism to function identically yet all nodes would benefit from this "full mempool scope" aggregation. Miners may even have an innate incentive to do so when the mempool is larger than block size, because they'd be able to gather more txn fees per block.

IIUC, if we *also* assume blocks contain *only* the coinbase and these aggregates, then this should notably improve validation time per block. Correct?

### Use Case Impact Brainstorm

In the case that this feature notably improves per-block validation time for full nodes, this could be a notable improvement for:

- Miners, where block validation delays present a substantial opportunity cost to PoW effort. Miners either must validate an incoming block fully, or begin performing PoW work on a block that is not yet fully validated, so lowering the validation time makes miners more resilient against the cost of invalid blocks.
- Low compute / high bandwidth scenarios such as for special purpose full node devices, point-of-sale kiosks, "full node in a box" products (ex: Casa, Nodl, …).
- Any "embedded consensus" clients, such as:
  - Ethereum smart contracts that replicate Zcash validation (ex: Keep tBTC does this for BTC).
  - Potentially embedded / nested ZKP proofs about Zcash blocks.
- Potentially "SPV-style" light wallets which are doing block validation.

### Further Improvements

There are some obvious further improvements which could build off of this feature:

- cross-block aggregation: There seems to be a spectrum between Zcash NU4 era design and fully succinct block chains. (This also seems to be the easiest next step in this list to me, with notable benefits to bootstrap/sync time and light wallet security.)
- nullifier / commitment aggregation, combined with a different approach to payment detection: if done well this would enable substantial capacity improvements that might look like ZKP roll-ups (which maintain strong privacy).
- sharding: I am increasingly curious about the relationship between sharding and ZKP roll-ups.

**Comment by daira** [comment]
> Is it possible for a miner to perform aggregation over their entire mempool?

Yes. It might be concretely quite expensive though. Typically miners want to include new transactions (or aggregates) in the template they're mining as soon as they have been validated, which is why I didn't assume that miners would do the aggregation.

**Comment by daira** [comment]
The original proposal had a bug — the recipient of a note found by scanning outputs would not have had sufficient information to construct its witness. The fix is to include commitments, which slightly reduces the bandwidth savings.

I have also added the distinction between non-trimmable and trimmable fields. Note that it is the total footprint that determines throughput, but trimming does help for node storage costs and catch-up bandwidth.

**Comment by daira** [comment]
If we wanted to retrofit the trimmable/non-trimmable distinction onto the existing shielded protocols, the fields that are trimmable in Sprout, Sapling, and Orchard, are the note ciphertexts and outgoing ciphertexts.

**Comment by daira** [comment]
@nathan-at-least wrote:
> sharding: I am increasingly curious about the relationship between sharding and ZKP roll-ups.

I think that the approach I outlined in my Zcon1 and Amsterdam talks, in which the nullifier set is sharded but there is a single address space and anonymity set, is the best way to do sharding for Zcash. But I'm also increasingly of the opinion that sharding is complex and more disruptive to the protocol than other scaling techniques, and so we should not do it unless and until it becomes clear that other techniques are not sufficient.

**Comment by kranzj** [comment]
This sounds massively cool :). Also one nice aspect of this is - correct me if I'm wrong due to my minimal understanding so far - that not all data will even be available on the blockchain. Thus, in case all security of Zcash falls apart one day, an attacker still lacks all puzzle pieces to build the entire transaction graph. I think having a P2P aspect of transacting is a so far underrated property of blockchain privacy.

Again, maybe I'm totally off, but it reminds me a small bit of Bitcoin's Cashu, like a trust-less decentralized version of it, without having to trust custodian (which totally ruins that idea).
